### DISCUSSION TITLE
useContractRead(s) in a for loop?

### DISCUSSION THREAD

**Discussion by nickyoung92** [wevm/wagmi#1534] [Q&A]
Trying to figure out how to iteratively call a contract read and push to array. Is there a better way? Is it currently possible? Tried this to no avail. 

```
function iterateRead() {
  var arr = []
  for(var index = 0; index < 5; i++) {
    var i = index + 1
    const contractDeployed = useContractRead({
      address: contractAddress,
      abi: contractABI,
      functionName: 'functionName',
      args: [address, i],
      onSuccess(data) {
        console.log(data)
        arr.push(data)
      },
    })
  }
}
```
function is called by button onClick and throws this error 
```
Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
```
I also tried useContractReads (which I am using perfectly fine in a separate instance) but I am not attempting to iterate that.

Any guidance is appreciated. Happy Holidevs

**Comment by tmm** [comment]
Check out [`useContractReads`](https://wagmi.sh/react/hooks/useContractReads).

**Reply by nickyoung92** [reply]
Would work if I new how many times to loop, but I dont. Could be 5, could be 3, could be 0, depending on another contract read.

**Reply by ECBSJ** [reply]
U ever figure out a way to use useContractRead(s) in a for loop yet? I'm also having the same issue...

**Reply by ECBSJ** [reply]
To work around this for now, you could just imperatively call (in your `for` loop) the contract method from the `ethers.Contract` instance returned from the `useContract` hook.

Just make sure you pass in a `signerOrProvider` option in the object.

**Reply by tmm** [reply]
Something like this:

```ts
import { useContractReads } from 'wagmi'

const contracts = React.memo(() => {
  // do looped logic in here
}, [deps])

useContractRead({ contracts })
```

**Reply by ECBSJ** [reply]
> Something like this:
> 
> ```ts
> import { useContractReads } from 'wagmi'
> 
> const contracts = React.memo(() => {
>   // do looped logic in here
> }, [deps])
> 
> useContractRead({ contracts })
> ```

Would you be able to expand the example you gave? I'm still trying to figure out how that would work.

**Reply by tmm** [reply]
```ts
import { useContractReads } from 'wagmi'

const contracts = React.memo(() => {
  return […].map((x) => ({
    address: x.address,
    abi: […],
    functionName: 'foo',
    args: […],
  })
}, [deps])

useContractRead({ contracts })
```

**Reply by ChrisChan888** [reply]
> ```ts
> import { useContractReads } from 'wagmi'
> 
> const contracts = React.memo(() => {
>   return […].map((x) => ({
>     address: x.address,
>     abi: […],
>     functionName: 'foo',
>     args: […],
>   })
> }, [deps])
> 
> useContractRead({ contracts })
> ```

Will contracts provide strongly typed in this case?

**Reply by tmm** [reply]
No, not unless you help out the compiler.
